### PR TITLE
[HTD-7] Feat: Add `Button` and `Dropdown` components

### PR DESCRIPTION
**Pull request by RBoraurito**
# Pull Request Description

## Description

Added reset css

## Assets

| before | after |
| ------ | ----- |
|        |       |

## Before Merge Checklist

The following steps were checked before merging a PR

- [x] checked and fixed code style issues
- [x] checked and removed unnecessary/junky code
- [x] manually tested introduced solution
- [ ] added needed tests to automate testing this code in the future

### CODE REVIEW

File: src/app/layout.tsx [hunk 1]
```diff
@@ -5,6 +5,8 @@ import { StyledLayout } from '@/components/layout/styled-layout';
 import { siteMetadata } from '@/lib/site-metadata';
 import StyledComponentsRegistry from '@/styles/styled-component-registry';
 
+import '@/styles/reset.css';
```
Comment (nephlin7) [inline src/app/layout.tsx]: @RBoraurito, currently, there is a reset in the `GlobalStyles` coming from `styled-reset`; could you delete it to use only `reset.css` instead?

File: src/app/page.tsx [hunk 2]
```diff
@@ -1,6 +1,12 @@
+'use client';
+
+import { CaretDown, RocketLaunch } from '@phosphor-icons/react/dist/ssr';
```
Comment (nephlin7) [inline src/app/page.tsx:1]: Could we create an Icon element and export the phosphor icon from there? So if, for some reason, the icon library changes, we will have to change it only in one place. Probably could be done in the Icon task, so just keep it in mind, please.

File: src/components/elements/button/button.styles.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
+'use client';
+
+import { keyframes } from 'styled-components';
```
Comment (nephlin7) [inline src/components/elements/button/button.styles.ts]: Let's import `keyframes` from `@/styles` please like css and styled

File: src/components/elements/button/button.styles.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    switch (props.variant) {
+      case 'primary':
+        return css`
+          color: ${({ theme }) => theme.colors.text.alwayWhite};
+          background: linear-gradient(${({ theme }) => theme.colors.orangeGradient});
+
+          border: 1px solid #ec502d;
+
+          box-shadow: 0px 2px 1px 0px #ffffff26 inset;
+          box-shadow: 0px -2px 2px 0px #0000000d inset;
```
Comment (nephlin7) [inline src/components/elements/button/button.styles.ts]: Can we use the rgba value instead, like `rgba(0, 0, 0, 0.5)`? The same in other places, please.

Comment (nephlin7) [inline src/components/elements/button/button.styles.ts]: By the way, do the `box-shadow: 0px -2px 2px 0px #0000000d inset` styles override the first one?, The same in other places, i.e:

Screenshot 2024-02-22 at 09 46 42

File: src/components/elements/button/index.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { StyledButton, StyledButtonWrapper } from './button.styles';
+import type { ButtonProps } from './button.types';
+import { MovingBorder } from '../moving-border';
+
+export const Button = async ({
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]: Curious, why do we need async here?

File: src/components/elements/button/index.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  type = 'button',
+  withBorder = false,
+  children,
+  as = 'button',
+  icon: Icon,
+  iconWeight = 'fill',
+  onClick,
+  href,
+}: Component<ButtonProps>) => {
+  const base = () => (
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]: Can we render it as JSX to make it more readable?
```suggestion
  const BaseButton = () => (
```
Then...

``` jsx
...
<BaseButton />
...
```

File: src/components/elements/button/index.tsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    <StyledButton as={as} variant={variant} small={small} type={type} onClick={onClick} href={href}>
+      {children}
+      {Icon && <Icon size={16} weight={iconWeight} />}
+    </StyledButton>
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]: Can we consider the case where the button is an `a` use `next/link` component for internal routes and external the regular `a` tag? Also, please add some basic accessibility like `role="button"`, etc.

File: src/components/elements/button/index.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  );
+
+  if (withBorder) {
+    return (
+      <StyledButtonWrapper>
+        <MovingBorder rx="30%" ry="30%" />
+        {base()}
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]:
```suggestion
        <BaseButton />
```

File: src/components/elements/button/index.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+      </StyledButtonWrapper>
+    );
+  }
+  return base();
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]:
```suggestion
  }
  
  return <BaseButton />;
```

File: src/components/modules/dropdown/dropdown.types.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import type { ButtonProps } from '@/components/elements/button/button.types';
+
+export interface DropdownProps extends Omit<ButtonProps, 'onClick' | 'href'> {
+  items: {
+    content: string;
+    as?: 'button' | 'a';
+    onClick?: () => void;
+    href?: string;
+  }[];
```
Comment (nephlin7) [inline src/components/modules/dropdown/dropdown.types.ts]: Can we create the interface of the item and use it here:
```suggestion
  items: DropdownItems[];
```

File: src/components/modules/dropdown/index.tsx [hunk 11]
```diff
@@ -0,0 +1,22 @@
+'use client';
+
+import { DropdownOption, DropdownOptionsWrapper, DropdownWrapper } from './dropdown.styles';
+import type { DropdownProps } from './dropdown.types';
+
+import { Button } from '@/components/elements/button';
+
+export const Dropdown = ({ children, items, ...buttonProps }: Component<DropdownProps>) => {
```
Comment (nephlin7) [inline src/components/modules/dropdown/index.tsx]: 1. Let's make sure to add some accessibility also here, please.
2. The dropdown component uses the caret icon by default but is not here.
3. There is an issue with the item's alignment when the text is large
4. It seems that some shadow or background doesn't follow the border radius of the button
Screenshot 2024-02-22 at 10 21 36

5. The component doesn't follow the design styles
Screenshot 2024-02-22 at 10 27 00

File: src/components/elements/button/button.styles.ts [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    from {
+      transform: rotate(0deg);
+    }
+
+    to {
+      transform: rotate(360deg);
+    }
+`;
+
+export const StyledButton = styled.button<ButtonProps>`
```
Comment (nephlin7) [inline src/components/elements/button/button.styles.ts]: Take a look at the design, please. It seems the styles are not correct in the implementation. 
_, i.e., The gradient is not in the correct direction._ This kind of details.

**Review by nephlin7 — CHANGES_REQUESTED** [review]
@RBoraurito, I left some suggestions. 

As a general recommendation for futures PR, let's try to always match the design please in the implementations.

File: src/components/elements/button/button.styles.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  padding: ${small => (small ? '20px 24px' : '16px 24px')};
+
+  border-radius: 12px;
+  display: inline-flex;
+  gap: 8px;
+  align-items: center;
+
+  ${props => {
+    switch (props.variant) {
+      case 'primary':
```
Comment (nephlin7) [inline src/components/elements/button/button.styles.ts:31]: For the primary version, it seems there are two variants with different hover/focus styles. I think it is not included here.

**Review by nephlin7 — CHANGES_REQUESTED** [review]

File: src/app/layout.tsx (hunk 1 above)
Comment (RBoraurito) [reply src/app/layout.tsx]: @nephlin7 wouldn't it be better to add it to the Global styles?

File: src/components/elements/button/button.styles.ts (hunk 4 above)
Comment (RBoraurito) [reply src/components/elements/button/button.styles.ts]: I'll fix this

File: src/app/layout.tsx (hunk 1 above)
Comment (nephlin7) [reply src/app/layout.tsx]: @RBoraurito, sure Let's do it in a `@/styles/reset.ts` and use it the `css` helper from styled-components and remove the styled-reset package, please.

File: src/app/page.tsx (hunk 2 above)
Comment (RBoraurito) [reply src/app/page.tsx:1]: @nephlin7 They are a lot but I can do it for some of them

Comment (nephlin7) [reply src/app/page.tsx:1]: @RBoraurito, there is no need to add for all, just for the ones we are using on the way

File: src/components/elements/button/button.styles.ts (hunk 13 above)
Comment (nephlin7) [reply src/components/elements/button/button.styles.ts:31]: @RBoraurito, the primary version with border still has some differences on how it looks and in its different states hover/focus
Figma
Screenshot 2024-02-22 at 16 18 40
implementation
Screenshot 2024-02-22 at 16 18 33

File: src/components/modules/dropdown/dropdown.styles.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+  &:hover,
+  &:focus {
+    background-color: ${({ theme }) => theme.colors.darkGrey};
+  }
+`;
+export const LinkDropdownOption = styled(Link)<DropdownOptionProps>`
+  ${baseOption}
+`;
+export const DropdownOption = styled.button<DropdownOptionProps>`
```
Comment (nephlin7) [inline src/components/modules/dropdown/dropdown.styles.ts]:
```suggestion
`;

export const DropdownOption = styled.button<DropdownOptionProps>`
```

File: src/components/elements/button/index.tsx [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+  if (withBorder) {
+    return (
+      <StyledButtonWrapper>
+        <MovingBorder rx="30%" ry="30%" />
+        <BaseButton />
+      </StyledButtonWrapper>
+    );
+  }
+  return <BaseButton />;
```
Comment (nephlin7) [inline src/components/elements/button/index.tsx]:
```suggestion
  }
  
  return <BaseButton />;
```

File: src/app/layout.tsx (hunk 1 above)
Comment (nephlin7) [reply src/app/layout.tsx]: @RBoraurito ??

File: src/components/elements/button/button.styles.ts [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+          &:active {
+            background: linear-gradient(345.32deg, #a32d13 10.67%, #dc3814 89.91%);
+            box-shadow:
+              0px 2px 1px 0px rgba(255, 255, 255, 0.15) inset,
+              0px 4px 24px 0px rgba(245, 109, 79, 0.15),
+              0px -2px 2px 0px rgba(0, 0, 0, 0.05) inset;
+          }
+        `;
+      case 'secondary':
```
Comment (nephlin7) [inline src/components/elements/button/button.styles.ts:79]: Check the design again, please. The initial border is not transparent. The border is missing in the focus state...

File: src/components/modules/dropdown/dropdown.styles.ts [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+import type { DropdownOptionProps } from './dropdown.types';
+
+import { css, styled } from '@/styles';
+
+const openDropdown = css`
+  padding: 4px;
+  max-height: 300px;
+`;
+
+export const DropdownOptionsWrapper = styled.div`
```
Comment (nephlin7) [inline src/components/modules/dropdown/dropdown.styles.ts:16]: Check the design, please. The border is missing and should be the same as the parent button; also, the focus of the parent button is different:

**Figma**
Screenshot 2024-02-22 at 15 56 23

**Implementation**

Screenshot 2024-02-22 at 16 00 49

File: src/components/elements/button/button.types.ts [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import type { Icon, IconWeight } from '@phosphor-icons/react';
+import type { AriaAttributes, AriaRole, HTMLAttributeAnchorTarget } from 'react';
+
+export interface ButtonProps extends AriaAttributes {
+  as?: 'button' | 'a' | 'Link';
```
Comment (nephlin7) [inline src/components/elements/button/button.types.ts]: I think it is better to have just  `button`| `link`
```suggestion
  as?: 'button' | 'link';
```
And handle, which should be rendered in the component internally for the link, if isExternalUrl _(the value contains `http/https`)_ render a tag if no the Link component. wdyt?

File: src/components/elements/button/button.types.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  role?: AriaRole;
+  target?: HTMLAttributeAnchorTarget;
+  variant?: 'primary' | 'secondary';
+  withBorder?: boolean;
+  small?: boolean;
+  type?: 'submit' | 'button' | 'reset';
+  onClick?: (event: MouseEvent) => void;
+  href?: string;
```
Comment (nephlin7) [inline src/components/elements/button/button.types.ts]: Can we make the type stronger here? If the `as` value is `link` href should be required.

**Review by nephlin7 — CHANGES_REQUESTED** [review]

File: src/components/modules/dropdown/index.tsx (hunk 11 above)
Comment (nephlin7) [reply src/components/modules/dropdown/index.tsx]: @RBoraurito there is still some inconsistency

**Pull request merged**